src/caches/beliefpropagationcache.jl

```diff
@@ -244,3 +262,33 @@ end
 function update_factor(bp_cache, vertex, factor)
   return update_factors(bp_cache, [vertex], ITensor[factor])
 end
+
+function vertex_norm(bp_cache::BeliefPropagationCache, pv::PartitionVertex)
+  incoming_mts = environment(bp_cache, [pv])
+  local_state = factor(bp_cache, pv)
+  return scalar(vcat(incoming_mts, local_state))
+end
+
+function edge_norm(bp_cache::BeliefPropagationCache, pe::PartitionEdge)
+  return scalar(vcat(message(bp_cache, pe), message(bp_cache, reverse(pe))))
+end
+
+function vertex_norms(bp_cache::BeliefPropagationCache, pvs::Vector)
+  return [vertex_norm(bp_cache, pv) for pv in pvs]
+end
+
+function vertex_norms(bp_cache::BeliefPropagationCache)
+  return vertex_norms(bp_cache, partitionvertices(partitioned_itensornetwork(bp_cache)))
+end
+
+function edge_norms(bp_cache::BeliefPropagationCache, pes::Vector)
+  return [edge_norm(bp_cache, pe) for pe in pes]
+end
+
+function edge_norms(bp_cache::BeliefPropagationCache)
+  return edge_norms(bp_cache, partitionedges(partitioned_itensornetwork(bp_cache)))
+end
+
+function scalar_factors(bp_cache::BeliefPropagationCache)
```

Comment:
I think a better name for this would be `scalar_factors_quotient` since it is outputting the factors of the numerator and denominator.